### PR TITLE
refactor to make superCellSize and frameSize independant

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- param-files: add `numFrameSlots` to set the number of particle slots per frame
- Refactor the PIConGPU extents for the random number factory used in PIConGPU.
  - each number of RNG states for each supercell is now dimensionless

# Update old setups:
  - add to `memory.param` after the definition of `SuperCellSize`

```C++
    /** number of slots for particles within a frame */
    static constexpr uint32_t numFrameSlots = pmacc::math::CT::volume<SuperCellSize>::type::value;
```

# Refactorings required if you develop extensions for PIConGPU

There are now `numFrameSlots` RNG states per supercell available. Instead of mapping the worker ID to a cell within the supercell you should calculate the RNG state cell this way:

```C++
// offset of the superCell (in cells, without any guards) to the
// origin of the local domain
DataSpace<simDim> const localSuperCellOffset = superCellIdx - mapper.getGuardingSuperCells();
auto rngOffset = DataSpace<simDim>::create(0);
rngOffset.x() = worker.getWorkerIdx();
auto numRNGsPerSuperCell = DataSpace<simDim>::create(1);
numRNGsPerSuperCell.x() = numFrameSlots;
rngHandle.init(localSuperCellOffset * numRNGsPerSuperCell + rngOffset);
```

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
CC-ing: @BrianMarre, an upcoming change you should be aware of.

**Comment by PrometheusPi** [comment]
@psychocoderHPC It looks like your code fails consistently with the KHI example cmake case 2 - which adds radiation.

File: include/picongpu/particles/collision/InterCollision.hpp [hunk 1]
```diff
@@ -146,9 +145,11 @@ namespace picongpu
...
                     // origin of the local domain
                     DataSpace<simDim> const localSuperCellOffset = superCellIdx - mapper.getGuardingSuperCells();
-                    rngHandle.init(
-                        localSuperCellOffset * SuperCellSize::toRT()
-                        + DataSpaceOperations<simDim>::template map<SuperCellSize>(worker.getWorkerIdx()));
+                    auto rngOffset = DataSpace<simDim>::create(0);
+                    rngOffset.x() = worker.getWorkerIdx();
+                    auto numRNGsPerSuperCell = DataSpace<simDim>::create(1);
+                    numRNGsPerSuperCell.x() = numFrameSlots;
+                    rngHandle.init(localSuperCellOffset * numRNGsPerSuperCell + rngOffset);
```
Comment (PrometheusPi) [inline include/picongpu/particles/collision/InterCollision.hpp:152]: here I need a bit more explanation later - but that's just my curiosity not a problem for the pull request

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 2]
```diff
@@ -227,21 +227,17 @@ namespace picongpu
             size_t size;
 
             ISAAC_NO_HOST_DEVICE_WARNING
-            ISAAC_HOST_DEVICE_INLINE ParticleIterator(
-                size_t size,
-                ParticlesBoxType pb,
-                FramePtr firstFrame,
-                int frameSize)
+            ISAAC_HOST_DEVICE_INLINE ParticleIterator(size_t size, ParticlesBoxType pb, FramePtr firstFrame)
```
Comment (PrometheusPi) [inline include/picongpu/plugins/IsaacPlugin.hpp:230]: why did you put this on a single line?

File: include/picongpu/plugins/common/openPMDVersion.def [hunk 3]
```diff
@@ -40,7 +40,7 @@ namespace picongpu
          *
          * @attention If the version is changed please update openPMDWriter::checkIOFileVersionRestartCompatibility().
          */
-        static constexpr int picongpuIOVersionMajor = 2;
+        static constexpr int picongpuIOVersionMajor = 3;
```
Comment (PrometheusPi) [inline include/picongpu/plugins/common/openPMDVersion.def:43]: Critical: why do you increase the IO version when you refactor frame size?

File: include/picongpu/plugins/openPMD/openPMDWriter.hpp [hunk 4]
```diff
@@ -860,11 +860,16 @@ make sure that environment variable OPENPMD_BP_BACKEND is not set to ADIOS1.
...
                 std::string datasetName = params->openPMDSeries->meshesPath() + name;
 
+                auto numRNGsPerSuperCell = DataSpace<simDim>::create(1);
+                numRNGsPerSuperCell.x() = numFrameSlots;
                 // rng states are always of the domain size therefore query sizes from domain information
                 const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
-                pmacc::math::UInt64<simDim> recordLocalSizeDims = subGrid.getLocalDomain().size;
-                pmacc::math::UInt64<simDim> recordOffsetDims = subGrid.getLocalDomain().offset;
-                pmacc::math::UInt64<simDim> recordGlobalSizeDims = subGrid.getGlobalDomain().size;
+                pmacc::math::UInt64<simDim> recordLocalSizeDims{
```
Comment (PrometheusPi) [inline include/picongpu/plugins/openPMD/openPMDWriter.hpp:867]: This now is `numRNGsPerSuperCell / SuperCellSize::toRT()` larger/smaller - is this intended @psychocoderHPC

Comment (PrometheusPi) [reply include/picongpu/plugins/openPMD/openPMDWriter.hpp:867]: same applies to all below

File: include/picongpu/plugins/radiation/executeParticleFilter.hpp [hunk 5]
```diff
@@ -38,12 +38,18 @@ namespace picongpu::plugins::radiation
...
     {
         constexpr bool hasRadiationFilter
             = pmacc::traits::HasIdentifier<typename T_Species::FrameType, radiationMask>::type::value;
 
         if constexpr(hasRadiationFilter)
-            particles::Manipulate<picongpu::plugins::radiation::RadiationParticleFilter, T_Species>{}(currentStep);
+        {
+            auto executeFilter
+                = particles::Manipulate<picongpu::plugins::radiation::RadiationParticleFilter, T_Species>{};
+            executeFilter(currentStep);
```
Comment (PrometheusPi) [inline include/picongpu/plugins/radiation/executeParticleFilter.hpp:52]: as info: just a restructuring - has nothing to do with objective of this pull request

File: include/pmacc/exec/KernelLauncher.tpp [hunk 6]
```diff
@@ -74,6 +74,13 @@ namespace pmacc::exec::detail
             , m_gridExtent(DataSpace<traits::GetNComponents<T_VectorGrid>::value>(gridExtent).toDim3())
             , m_blockExtent(DataSpace<traits::GetNComponents<T_VectorBlock>::value>(blockExtent).toDim3())
         {
+#if 0
+            std::string const kernelName = typeid(m_kernel).name();
+            std::string const kernelInfo = kernelName + std::string(" [") + m_metaData.getFile() + std::string(":")
```
Comment (PrometheusPi) [inline include/pmacc/exec/KernelLauncher.tpp]: there is still not used code here

File: include/pmacc/particles/ParticlesBase.hpp [hunk 7]
```diff
@@ -71,12 +71,7 @@ namespace pmacc
         /* Policies for handling particles in guard cells */
         using HandleGuardRegion = typename ParticleDescription::HandleGuardRegion;
 
-        enum
-        {
-            Dim = MappingDesc::Dim,
-            Exchanges = traits::NumberOfExchanges<Dim>::value,
-            TileSize = math::CT::volume<typename MappingDesc::SuperCellSize>::type::value
-        };
```
Comment (PrometheusPi) [inline include/pmacc/particles/ParticlesBase.hpp:79]: Was this `enum` not used before?

File: oft
Comment (PrometheusPi) [inline oft:1]: critical: this should not be part of this pull request

File: share/picongpu/examples/FoilLCT/include/picongpu/param/memory.param [hunk 8]
```diff
@@ -49,6 +49,7 @@ namespace picongpu
      * volume of a superCell must be <= 1024
      */
     using SuperCellSize = mCT::Int<16, 16>;
+    static constexpr uint32_t numFrameSlots = pmacc::math::CT::volume<SuperCellSize>::type::value;
```
Comment (PrometheusPi) [inline share/picongpu/examples/FoilLCT/include/picongpu/param/memory.param:54]: the rest of `memeory.param` has a comment for this new line of code. This one doesn't.

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
A few questions and same stuff that needs to be removed/changed.

File: include/picongpu/plugins/common/openPMDVersion.def (hunk 3 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/common/openPMDVersion.def:43]: The way how the random number generator is stored has changed. Therefore old checkpoints are not compatible anymore. By increasing the version the user will get a incompatibility message in case he/she is using a old file.

File: include/picongpu/particles/collision/InterCollision.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/collision/InterCollision.hpp:152]: The storage format of the RNG states changed. In the past, we used one RNG state per cell. This is changes within this PR where we have a linear array of RNG states per superCell but stored in an array [x-superCells*RNG-states-perSuperCell,y-supercell,z-supercell].

This will be implemented in the future a little bit more cleanly so that we really work with a hierarchy supercell equal to the particle hierarchy structure.

File: include/picongpu/plugins/openPMD/openPMDWriter.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/openPMD/openPMDWriter.hpp:867]: This is intended, we have now N-RNG states per supercell and not supercell cells RNG states per supercell.
N-is the number of slots in a frame because the number of slots is >= number of cells per supercell

File: include/picongpu/plugins/radiation/executeParticleFilter.hpp (hunk 5 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/radiation/executeParticleFilter.hpp:52]: After this PR the compiler is throwing a warning at this point. IMO the warning is a false positive. Therefore this tiny change is required in this PR.

File: include/pmacc/exec/KernelLauncher.tpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/pmacc/exec/KernelLauncher.tpp]: good catch, this is a left over which should be removed.

File: include/pmacc/particles/ParticlesBase.hpp (hunk 7 above)
Comment (psychocoderHPC) [reply include/pmacc/particles/ParticlesBase.hpp:79]: Yes some of this enums was not used. There will be a follow up PR where more of the enum stuff will be cleaned

File: share/picongpu/examples/FoilLCT/include/picongpu/param/memory.param (hunk 8 above)
Comment (psychocoderHPC) [reply share/picongpu/examples/FoilLCT/include/picongpu/param/memory.param:54]: fixed

File: include/pmacc/exec/KernelLauncher.tpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/pmacc/exec/KernelLauncher.tpp]: done

File: oft
Comment (PrometheusPi) [inline oft:1]: This file is still here

**Review by PrometheusPi — DISMISSED** [review]
The `oft` file is still in the pull request. Please remove.

File: include/picongpu/plugins/IsaacPlugin.hpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/IsaacPlugin.hpp:230]: This is auto-formated I cannot influence the style.

File: oft
Comment (psychocoderHPC) [reply oft:1]: ^^ will remove

**Comment by psychocoderHPC** [comment]
I will merge this PR to avoid that new code in other PR's is introducing wrong usage of cellsize and framesize
